spec/models/dynamic_field_group_spec.rb

```diff
@@ -0,0 +1,339 @@
+            parent.reload
+          end
+
+          it 'sets sort order to one more than the higest sort order of any sibling' do
+            expect(dynamic_field_group.sort_order).to be 17
+          end
+        end
+
+        context 'and no other sublings are present' do
+          it 'sets sort order to 0' do
+            expect(dynamic_field_group.sort_order).to be 0
+          end
+        end
+      end
+    end
+
+    context 'when created_by is missing' do
+      let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, created_by: nil) }
+
+      it 'does not save' do
+        expect(dynamic_field_group.save).to be false
+      end
+
+      it 'returns correct error' do
+        dynamic_field_group.save
+        expect(dynamic_field_group.errors.full_messages).to include 'Created by must exist'
+      end
+    end
+
+    context 'when updated_by is missing' do
+      let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, updated_by: nil) }
+
+      it 'does not save' do
+        expect(dynamic_field_group.save).to be false
+      end
+
+      it 'returns correct error' do
+        dynamic_field_group.save
+        expect(dynamic_field_group.errors.full_messages).to include 'Updated by must exist'
+      end
+    end
+
+    context 'when xml_translation is invalid' do
+      let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, xml_translation: 'randomstring') }
+
+      it 'does not save' do
+        expect(dynamic_field_group.save).to be false
+      end
+
+      it 'returns correct error' do
+        dynamic_field_group.save
+        expect(dynamic_field_group.errors.full_messages).to include 'Xml translation does not validate as JSON'
+      end
+    end
+
+    context 'when creating a group with the same name as a sibling' do
+      context 'and the sibiling is a dynamic_field' do
+        let(:parent) { FactoryBot.create(:dynamic_field_group) }
+        let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, :child, parent: parent) }
+
+        before do
+          FactoryBot.create(:dynamic_field, string_key: 'role', dynamic_field_group: parent)
+          parent.reload
+        end
+
+        it 'does not save' do
+          expect(dynamic_field_group.save).to be false
+        end
+
+        it 'returns correct error' do
+          dynamic_field_group.save
+          expect(dynamic_field_group.errors.full_messages).to include 'String key has already been taken'
+        end
+      end
+
+      context 'and the sibiling is a dynamic_field_group' do
+        let(:parent) { FactoryBot.create(:dynamic_field_category) }
+        let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, parent: parent) }
+
+        before do
+          FactoryBot.create(:dynamic_field_group, string_key: 'name', parent: parent)
+          parent.reload
+        end
+
+        it 'does not save' do
+          expect(dynamic_field_group.save).to be false
```

Comment:
The only thing I don't like about this test is that it assumes knowledge of the factory.  It assumes that you know that the factory for :dynamic_field_group defaults to having a string_key of 'name'.  Maybe that's an okay practice though.  Not sure, since I only recently started using FactoryBot.

Reply:
I am not sure what the practice should be in cases like this. I just don't want to keep writing boiler plate models over and over again, so I depend on the string_key of the default factory. Might be worth aliasing the default factory to a trait called :name.